helion/_compiler/host_function.py

```diff
@@ -96,15 +96,34 @@ def __init__(
             self.args: ast.arguments = root.args
             self.body: list[ast.stmt] = root.body
 
+            HostFunction.validate_ast(root)
+
             from .device_ir import lower_to_device_ir
             from .type_propagation import propagate_types
 
             propagate_types(self, fake_args)
             env.errors.raise_if_errors()
             env.finalize_config_spec()
             self.device_ir = lower_to_device_ir(self)
-            # TODO(jansel): assert we don't have any extra decorators
-            # TODO(jansel): check type annotations for hl.constexpr/hl.specialize
+
+    @staticmethod
+    def validate_ast(root: ast.FunctionDef) -> None:
+        # Decorators are allowed before the helion kernel decorator
+        # but are not allowed after
+        def get_decorator_name(decorator: ast.expr) -> str:
+            if isinstance(decorator, ast.Name):
+                return decorator.id
+            if isinstance(decorator, ast.Attribute):
+                return get_decorator_name(decorator.value)
+            if isinstance(decorator, ast.Call):
+                return get_decorator_name(decorator.func)
+            raise AssertionError(f"Unknown decorator: {decorator}")
+
+        for idx, decorator in enumerate(root.decorator_list):
+            if get_decorator_name(decorator) == "helion":
+                assert idx == len(root.decorator_list) - 1, (
+                    "Decorators after helion kernel decorator are not allowed"
+                )
```

Comment:
Can you add a exception to `helion.exc` and raise that rather than asserting?

Asserts can be disabled by cpython flags and give worse error messages.